src/helpers/clipboard/clipboard_figure_state.ts

```diff
@@ -84,12 +84,20 @@ export class ClipboardFigureState implements ClipboardState {
    */
   paste(target: Zone[]) {
     const sheetId = this.getters.getActiveSheetId();
+    const size = { height: this.copiedFigure.height, width: this.copiedFigure.width };
+    const viewportRect = this.getters.getMainViewportRect();
+    const colDimensions = this.getters.getColDimensions(sheetId, target[0].left);
+    const rowDimensions = this.getters.getRowDimensions(sheetId, target[0].top);
     const position = {
-      x: this.getters.getColDimensions(sheetId, target[0].left).start,
-      y: this.getters.getRowDimensions(sheetId, target[0].top).start,
+      x:
+        viewportRect.width - colDimensions.start < size.width
+          ? viewportRect.width + viewportRect.x - size.width - 96
```

Comment:
also don't use random numbers in the operations. What are 96 and 28?

Reply:
Update the code and use the dimension of last column and row for identifying the end coordinate of the sheet